### ISSUE TITLE
1819 Sem 2 - Comparators

### ISSUE THREAD

**Issue by chamath2030** [nus-cs2030/2021-s2#302]
# Summary
Not sure if I did this question correctly, particularly Part C, where we have to design a student comparator class. Please advise

## Description 
For part A, I wrote
`Comparator<Integer> comp1 = (x, y) -> x - y; 
Comparator<String> comp2 = (x, y) -> x.compareTo(y);`

for part B, 
`List<Comparator> comparatorList = new ArrayList<Comparator>();
comparatorList.add(comp1);
comparatorList.add(comp2);`

## Suggested Answer (if any):
for part C, 
`class StudentComparator implements Comparator<T> {
   public int compare(T o1, T o2) {
   return o1 - o2;
   }
 }`

## Screenshots (if any): 
![image](https://user-images.githubusercontent.com/82226334/115695753-a5f95400-a394-11eb-9a5f-c12d06b0b1e6.png)

**Comment by ShyamalSankar** [comment]
Hello, I think the comparators have to be Comparator<Student>. So for the first example it would be something like
Comparator<Student> comp1 = (x,y) -> x.getLabID() - y.getLabID();

As for part 2 I think the list should contain Comparator<Student> so List<Comparator<Student>>.


Part C I am not a 100 percent sure but i created a class that takes in a list of comparators that compares based on the first comparator of the list and breaks ties by using the second comparator on the list.

**Comment by Onelayer** [comment]
Hello,

For part A, you want the comparators to compare the labID and the tutID of the students. As such, the comparator will take in two students to compare. Hence instead of using compareTo on x & y, you should instead use x.getLabID().compareTo(y.getLabID()) instead as the Integer class has a compareTo method which it inherit from the Comparable interface. 

For part C, my StudentComparator takes in a list as an attribute to store the Comparators. The compare method then uses the first the two comparators in order to see whether to return -1, 0 or 1. It should use the 2nd Comparator as the tie breaker.

**Comment by eugenelimnus** [comment]
Hi, for part (c) i found the thenComparing(Comparator<? super T> other) method in Comparator quite useful (not 100% sure if this is the correct usage), it returns a lexicographic-order comparator with another comparator, here is my code, though the question might be asking us to iterate through the list of comparators (which i didnt do).
![image](https://user-images.githubusercontent.com/77771438/115978761-7759c400-a5b4-11eb-9500-86fa306d3971.png)

**Comment by zhiqixn** [comment]
Hi, just to clarify, the class Student Comparator is only responsible for returning the results of a comparison between two students? So we just need to implement a compare() method that returns an integer output and not the entire student process right? I was a bit confused by the question

**Comment by yaohua98** [comment]
> Hi, just to clarify, the class Student Comparator is only responsible for returning the results of a comparison between two students? So we just need to implement a compare() method that returns an integer output and not the entire student process right? I was a bit confused by the question
> 
> Also, not very sure about how to answer part (d)? Are the four prompts what we are expected to print out?

Hi my dear Thearts head ,
the question states "We can achieve this by storing the comparators (one
or more) in a list, and process the comparators one by one when sorting students."
I believe the implementation of the sorting process only needs to be shown in part c/ d.
For part d, those 4 prompts are just examples of students imo!

**Comment by e0588269** [comment]
Hello! Im still not sure how to answer part(d). Do we have to create an ArrayList to take in the 4 students and then use the appropriate comparators on them? Is the use of bubble sort/selection sort necessary here?

**Comment by gracefour11** [comment]
yup, the question is essentially asking how to test ur program with the `StudentComparator` you've created. 4 students are given and they would need to be sorted accordingly. if you're gonna use `ArrayList`, u can use the `Collections.sort `to sort the students.

**Comment by dannytayjy** [comment]
For part d, you will need to store all the students in an `List<student> studentList`, after which use `Collections.sort(studentList, new StudentComparator(comparatorList))`.

**Comment by ashleyleerx** [comment]
For part (c) I created the Student Comparator class which takes in the two mentioned comparators and I generated a new comparator using the 2 comparators i.e   this.comp = first.thenComparing(second); before overriding the compare method
